<a href="https://colab.research.google.com/github/arashakbari1234/Data-Mining/blob/main/Assignment_3_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.datasets import fetch_lfw_people
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report, plot_confusion_matrix, accuracy_score
from sklearn.decomposition import PCA
from sklearn.neural_network import MLPClassifier

In [ ]:
lfw_people = fetch_lfw_people(min_faces_per_person=150, resize=0.4)
X = lfw_people.data
y = lfw_people.target

In [ ]:
print(lfw_people.data.shape)
print(lfw_people.target.shape)
Name = lfw_people.target_names

print(Name)

(766, 1850)
(766,)
['Colin Powell' 'George W Bush']


In [ ]:
pca_model = PCA(n_components=50)
X = pca_model.fit_transform(X)
pca_model.components_
np.sum(pca_model.explained_variance_ratio_)

0.87171644

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=101, shuffle=True)

In [ ]:
print(X_train.shape)
print(y_train.shape)
print(X_train[0].shape)

(574, 50)
(574,)
(50,)


In [ ]:
clf = MLPClassifier(hidden_layer_sizes=(1024,), batch_size=256, verbose=False, early_stopping=True)

In [ ]:
clf.fit(X_train, y_train)

MLPClassifier(batch_size=256, early_stopping=True, hidden_layer_sizes=(1024,))

In [ ]:
y_pred = clf.predict(X_test)

In [ ]:
# pred = np.argmax(y_pred,axis=1)
# ground=np.argmax(testy,axis=1)
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.83      0.80      0.82        61
           1       0.91      0.92      0.92       131

    accuracy                           0.89       192
   macro avg       0.87      0.86      0.87       192
weighted avg       0.88      0.89      0.88       192



In [ ]:
y_t = clf.predict(X_train)
print(classification_report(y_train,y_t))

              precision    recall  f1-score   support

           0       0.93      0.85      0.89       175
           1       0.94      0.97      0.95       399

    accuracy                           0.94       574
   macro avg       0.93      0.91      0.92       574
weighted avg       0.94      0.94      0.93       574



As we can see, we got training accuracy of 94% and test accuracy of 89% and clearly doesn't overfit.
we used PCA and earlyStoping and L2 regularization to avoid overfit in this problem.
Using PCA and MLP is a common technique for different problems with high dimentional features and as we can see the result is quite accurate and reliable. Deciding between the recall, precision and f1-score depend on the problem and trying to understand which one is more important to us.